ant/easy/channel.py

```diff
@@ -113,6 +118,26 @@ def send_broadcast_data(self, data):
         self._ant.send_broadcast_data(self.id, data)
 
     def send_acknowledged_data(self, data):
+        # SDS WORKAROUND : eerst kijken of nog uitstaande TX_COMPLETED events in de node._events zitten voor dit channel
+        # voor het geval er nog een is aangekomen buiten timeout van een vorige send_acknowledged_data
+        # momenteel verhindert niets om meerdere send_acknowledged_data op hetzelfde channel terzelfdertijd te starten vanuit <> threads
+        # dus deze workaround kan deze events wel onterecht wissen!
+        self._node._event_cond.acquire()
+        msgs_to_delete = []
+        for message in self._node._events:
+            if (message[0], message[1], message[2][0]) == (
+                self.id,
+                1,
+                Message.Code.EVENT_TRANSFER_TX_COMPLETED,
+            ):
+                msgs_to_delete.append(message)
+
+        for message in msgs_to_delete:
+            self._node._events.remove(message)
+
+        self._node._event_cond.release()
```

Comment:
Could you explain this bit? My Dutch is a bit rusty :).

Reply:
Haha! First of all, I apologize for the dutch, should have cleaned that. If a TX_COMPLETED event arrives after the timeout, the event will end up in the node._events queue, and a subsequent send_acknowledged_data will look in the queue, find a TX_COMPLETED event, but not the right one, and unblock prematurely. The cause is the wait_for_message function, that doesn't match events with the channel; and doesn' guarantee a minimum timeout : for _ in range(10): when a burst of 10 events arrives, even on a different channel, the timeout is quasi nihil. In my case the events were especially QUE_OVERFLOW and EVENT_QUE_OVERFLOW that flooded in, until I removed the sleep(0.1) line in ant.py and the usb-read thread could handle all the incoming data.
The solution needs rework on different files, and haven't gotten there yet. The workaround simply checks if there are trailing TRANSFER_TX_COMPLETED events in the queue, and deletes them.